pkg/porter/reconcile.go

```diff
@@ -0,0 +1,205 @@
+	// Have we successfully completed the install action?
+	if !i.Status.InstallationCompleted || lastRun == nil {
+		fmt.Fprintln(p.Err, "Triggering because the installation has not completed successfully yet")
+		return false, nil
+	}
+
+	opts := action.GetOptions()
+
+	newRef, err := p.resolveBundleReference(opts)
+	if err != nil {
+		return false, err
+	}
+
+	// Has the bundle definition changed?
+	if lastRun.BundleDigest != newRef.Digest.String() {
+		fmt.Fprintf(p.Err, "Triggering because the bundle definition has changed from %s(%s) to %s(%s)\n", lastRun.BundleReference, lastRun.BundleDigest, newRef.Reference, newRef.Digest)
+		return false, nil
+	}
+
+	// Has the bundle parameters changed?
+	if err := opts.LoadParameters(p); err != nil {
+		return false, err
+	}
+
+	// Get a set of parameters ready for comparison to another set of parameters
+	// to tell if the installation should be executed again. For now I'm just
+	// removing internal parameters (e.g. porter-debug, porter-state) and making
+	// sure that the types are correct, etc.
+	b := newRef.Definition
+	resolvedParams, err := p.resolveParameters(i, b, action.GetAction(), opts.combinedParameters)
+	if err != nil {
+		return false, err
+	}
+
+	// Convert parameters to a string to compare them. This avoids problems comparing
+	// values that may be equal but have different types due to how the parameter
+	// value was loaded.
+	prepParametersForComparison := func(params map[string]interface{}) (map[string]string, error) {
+		compParams := make(map[string]string, len(params))
+		for paramName, rawValue := range params {
+			if b.IsInternalParameter(paramName) {
+				continue
+			}
+
+			typedValue, err := b.ConvertParameterValue(paramName, rawValue)
+			if err != nil {
+				return nil, err
+			}
+
+			stringValue, err := b.WriteParameterToString(paramName, typedValue)
+			if err != nil {
+				return nil, err
+			}
+
+			compParams[paramName] = stringValue
+		}
+		return compParams, nil
+	}
+
+	oldParams, err := prepParametersForComparison(lastRun.Parameters)
+	if err != nil {
+		return false, errors.Wrapf(err, "error prepping previous parameters for comparision")
+	}
+
+	newParams, err := prepParametersForComparison(resolvedParams)
+	if err != nil {
+		return false, errors.Wrapf(err, "error prepping current parameters for comparision")
+	}
+
+	if !cmp.Equal(oldParams, newParams) {
+		diff := cmp.Diff(oldParams, newParams)
+		fmt.Fprintf(p.Err, "Triggering because the parameters have changed.\nDiff:\n%s\n", diff)
+		return false, nil
+	}
+
+	// Check only if the names of the associated credential sets have changed
+	// This is a "good enough for now" decision that can be revisited if we
+	// get use cases for needing to diff the actual credentials.
+	sort.Strings(lastRun.CredentialSets)
+	sort.Strings(i.CredentialSets)
+	if !cmp.Equal(lastRun.CredentialSets, i.CredentialSets) {
```

Comment:
https://github.com/google/go-cmp is nice because you can define custom equality functions, it provides a diff of the values, and overall an improvement over DeepEqual. I mostly like the diff output 😀

Reply:
It was also already in our dependency graph too, so it was an easy choice to pull it in.